crucible-mir-comp/src/Mir/Compositional/Convert.hs

```diff
@@ -306,6 +308,86 @@ exprToTerm sym sc w4VarMapRef val = liftIO $ do
     term <- SAW.scVariable sc ec
     return term
 
+
+-- | Try to convert a Crucible register value into a SAW core `Term`, using
+-- a `CryTermAdaptor` to validate and convert references to slices.
+-- We check that references to slices match the expected length specified
+-- in the `CryTermAdaptor`, and if so we read out the value stored in
+-- in the slice.
+regToTermWithAdapt :: forall m p sym t fs tp0 rtp args ret.
+    ( IsSymInterface sym,
+      sym ~ MirSym t fs,
+      m ~ OverrideSim (p sym) sym MIR rtp args ret) =>
+    sym ->
+    SAW.SharedContext ->
+    String ->
+    IORef (Map SAW.VarIndex (Some (W4.Expr t))) ->
+    CryTermAdaptor Integer ->
+    TypeShape tp0 ->
+    RegValue sym tp0 ->
+    m SAW.Term
+regToTermWithAdapt sym sc name w4VarMapRef ada0 shp0 rv0 = go ada0 shp0 rv0
+  where
+    go :: forall tp.
+        CryTermAdaptor Integer ->
+        TypeShape tp ->
+        RegValue sym tp ->
+        m SAW.Term
+    go ada shp rv =
+      case (ada, shp, rv) of
+        (NoAdapt, _, _) -> regToTerm sym sc name w4VarMapRef shp rv
+        (AdaptTuple as, TupleShape _ elems, ag) -> do
+            terms <- accessMirAggregate' sym elems as ag $ \_off _sz shp' rv' a -> go a shp' rv'
+            liftIO $ SAW.scTuple sc terms
+        (AdaptArray a, ArrayShape _ _ shp', vec) -> do
+            terms <- goVector a shp' vec
+            tyTerm <- shapeToTerm' sc a shp'
+            liftIO $ SAW.scVector sc tyTerm terms
+        (AdaptDerefSlice n, SliceShape _ty elT M.Immut tpr, Ctx.Empty Ctx.:> RV mirPtr Ctx.:> RV lenExpr) ->
+          case BV.asUnsigned <$> W4.asBV lenExpr of
+            Nothing ->
+              fail "Slice length is not statically known"
+
+            Just n1
+              | AsBaseType baseT <- asBaseType tpr ->
```

Comment:
Echoing my comments in https://github.com/GaloisInc/saw-script/pull/2653#discussion_r2383671954, I'm still skeptical about the use of `asBaseType` here. I think we could straightforwardly generalize this to work over any `TypeShape` by calling [`tyToShapeEq`](https://github.com/GaloisInc/saw-script/blob/5f6dca23315ad19d189d305db81c3437bd600c0a/saw-central/src/SAWCentral/Crucible/MIR/TypeShape.hs#L328-L337) to convert `tpr` from a `TypeRepr` to a `TypeShape`. (We'd have to pass in a `Mir.Mir.Collection` in order to do so, but this can be obtained nearby from the `loadCryptolFunc` function, which calls `regToTermWithAdapt`.)